locmap/lib/deletion.js

```diff
@@ -0,0 +1,120 @@
+'use strict';
+
+/*
+    Deleting *all* data related to a user from the database.
+*/
+
+var db = require('../../lib/db');
+var logger = require('../../lib/logger');
+var conf = require('../../lib/config');
+var LocMapUserModel = require('./locMapUserModel');
+var LocMapShareModel = require('./locMapShareModel');
+
+var assert = require('assert');
+var suspend = require('suspend');
+
+var userPrefix = conf.get('db').userPrefix;
+var sharePrefix = conf.get('db').sharePrefix;
+var codePrefix = conf.get('db').deleteCodePrefix; 
+
+exports.deleteUser = suspend(function* (userId, callback) {
+
+    try {
+        yield db.del(userPrefix + userId, suspend.resume());
+        yield db.del(sharePrefix + userId, suspend.resume());
+        yield db.del(codePrefix + userId, suspend.resume());
+
+        var shareKeys = yield db.keys(sharePrefix+'*', suspend.resume());
+
+        for (var i = 0; i < shareKeys.length; ++i) {
+
+            var id = shareKeys[i].split(':')[1];
+            var share = new LocMapShareModel(id);
+
+            var result = yield share.getData(suspend.resumeRaw());
+            
+            if (result[0] === 404) {
+                throw new Error('Could not read data for '+shareKeys[i]);
+            } else {
+                var data = result[0];
+            }
+
+            var lists = [
+                data.canSeeMe, 
+                data.ICanSee, 
+                data.ignored
+            ];
+            
+            for (var j = 0; j < lists.length; ++j) {
+                var ix = lists[j].indexOf(userId);
+                if (j > -1) {
+                    lists[j].splice(ix, 1);
+                }
+            }
+
+            if (!!data.nameMapping) {
+                var mapping = JSON.parse(data.nameMapping);
+                if (!!mapping[userId]) {
+                    delete mapping[userId];
```

Comment:
delete doesn't remove element from array, but sets it undefined instread.  You might want to use splice here.